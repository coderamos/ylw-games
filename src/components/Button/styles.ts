import styled, { css, DefaultTheme } from 'styled-components';
import { darken } from 'polished';

import { ButtonProps } from '.';

export type WrapperProps = {
  hasIcon: boolean;
} & Pick<ButtonProps, 'size' | 'fullWidth' | 'minimal'>;

const buttonModifiers = {
  small: (theme: DefaultTheme) => css`
    font-size: ${theme.font.sizes.xsmall};
    height: 3rem;
  `,
  medium: (theme: DefaultTheme) => css`
    font-size: ${theme.font.sizes.small};
    height: 4rem;
    padding: ${theme.spacings.xxsmall} ${theme.spacings.medium};
  `,
  large: (theme: DefaultTheme) => css`
    font-size: ${theme.font.sizes.medium};
    height: 5rem;
    padding: ${theme.spacings.xxsmall} ${theme.spacings.xlarge};
  `,
  fullWidth: () => css`
    width: 100%;
  `,
  withIcon: (theme: DefaultTheme) => css`
    svg {
      width: 1.5rem;
      & + span {
        margin-left: ${theme.spacings.xxsmall};
      }
    }
  `,
  minimal: (theme: DefaultTheme) => css`
    background: none;
    color: ${theme.colors.primary};

    &:hover {
      color: ${darken(0.1, theme.colors.primary)};
    }
  `
};

export const Wrapper = styled.button<WrapperProps>`
  ${({ theme, size, fullWidth, hasIcon, minimal }) => css`
    align-items: center;
    display: inline-flex;
    justify-content: center;
    background: ${theme.colors.primary};
    border-radius: ${theme.border.radius};
    border: none;
    color: ${theme.colors.secondary};
    cursor: pointer;
    padding: ${theme.spacings.xxsmall};
    transition: background 0.3s;
    text-decoration: none;

    &:hover {
      background: ${minimal ? 'none' : theme.colors.primaryHover};
    }

    ${!!size && buttonModifiers[size](theme)};
    ${!!fullWidth && buttonModifiers.fullWidth};
    ${!!hasIcon && buttonModifiers.withIcon(theme)};
    ${!!minimal && buttonModifiers.minimal(theme)};
  `}
`;

export const IconWrapper = styled.span``;
