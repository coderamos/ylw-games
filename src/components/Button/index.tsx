import { AnchorHTMLAttributes, ButtonHTMLAttributes } from 'react';
import * as s from './styles';

type ButtonTypes =
  | AnchorHTMLAttributes<HTMLAnchorElement>
  | ButtonHTMLAttributes<HTMLButtonElement>;

export type ButtonProps = {
  size?: 'small' | 'medium' | 'large';
  fullWidth?: boolean;
  minimal?: boolean;
  icon?: React.ReactNode;
  as?: 'button' | 'a';
} & ButtonTypes;

const Button = ({
  children,
  fullWidth = false,
  minimal = false,
  icon,
  size = 'medium',
  ...rest
}: ButtonProps) => (
  <s.Container
    size={size}
    fullWidth={fullWidth}
    hasIcon={!!icon}
    minimal={minimal}
    {...rest}
  >
    {!!icon && icon}
    {!!children && <s.IconWrapper>{children}</s.IconWrapper>}
  </s.Container>
);

export default Button;
