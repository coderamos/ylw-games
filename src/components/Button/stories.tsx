import { Story, Meta } from '@storybook/react/types-6-0';
import { ShoppingCart as ShoppingCartIcon } from '@styled-icons/feather/ShoppingCart';
import Button from '.';

export default {
  title: 'Design Stystem/Button',
  component: Button,
  argTypes: {
    children: {
      type: 'string'
    },
    icon: {
      type: ''
    }
  }
} as Meta;

export const Default: Story = args => <Button {...args} />;

Default.args = {
  children: 'Button',
  size: 'medium',
  fullWidth: false
};

export const ButtonWithIcon: Story = args => <Button {...args} />;

ButtonWithIcon.args = {
  children: 'Button',
  icon: <ShoppingCartIcon />,
  size: 'medium'
};

export const ButtonAsLink: Story = args => <Button {...args} />;

ButtonAsLink.args = {
  children: 'Button',
  as: 'a',
  href: '/link'
};
