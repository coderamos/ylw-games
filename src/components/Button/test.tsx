import { screen } from '@testing-library/react';
import { renderWithTheme } from 'utils/tests/helpers';
import { ShoppingCart as ShoppingCartIcon } from '@styled-icons/feather/ShoppingCart';

import Button from '.';

describe('<Button />', () => {
  it('should render the medium size button by default', () => {
    const { container } = renderWithTheme(<Button>Button</Button>);

    expect(screen.getByRole('button', { name: /Button/i })).toHaveStyle({
      'font-size': '1.4rem',
      height: '4rem',
      padding: '0.8rem 3.2rem'
    });

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render the small size button', () => {
    renderWithTheme(<Button size="small">Button</Button>);

    expect(screen.getByRole('button', { name: /Button/i })).toHaveStyle({
      'font-size': '1.2rem',
      height: '3rem'
    });
  });

  it('should render the large size button', () => {
    renderWithTheme(<Button size="large">Button</Button>);

    expect(screen.getByRole('button', { name: /Button/i })).toHaveStyle({
      'font-size': '1.6rem',
      height: '5rem',
      padding: '0.8rem 4.8rem'
    });
  });

  it('should render full width button', () => {
    renderWithTheme(<Button fullWidth>Button</Button>);

    expect(screen.getByRole('button', { name: /Button/i })).toHaveStyle({
      width: '100%'
    });
  });

  it('should render button with icon based in prop', () => {
    renderWithTheme(
      <Button icon={<ShoppingCartIcon data-testid="icon" />}>Button</Button>
    );

    expect(screen.getByText(/Button/i)).toBeInTheDocument();
    expect(screen.getByTestId('icon')).toBeInTheDocument();
  });

  it('should render a minimal version based in prop', () => {
    renderWithTheme(
      <Button icon={<ShoppingCartIcon data-testid="icon" />} minimal>
        Button
      </Button>
    );

    expect(screen.getByRole('button', { name: /Button/i })).toHaveStyle({
      background: 'none',
      color: '#FFCB00'
    });

    expect(screen.getByRole('button', { name: /Button/i })).toHaveStyleRule(
      'background',
      'none',
      {
        modifier: ':hover'
      }
    );
  });

  it('should render Button as a link', () => {
    renderWithTheme(
      <Button as="a" href="/link">
        Button
      </Button>
    );

    expect(screen.getByRole('link', { name: /button/i })).toHaveAttribute(
      'href',
      '/link'
    );
  });
});
