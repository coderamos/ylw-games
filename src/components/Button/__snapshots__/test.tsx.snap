// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should render the medium size button by default 1`] = `
.c0 {
  background: #FFCB00;
  border-radius: 0.4rem;
  border: none;
  color: #171717;
  cursor: pointer;
  padding: 0.8rem;
  -webkit-transition: background 0.3s;
  transition: background 0.3s;
  font-size: 1.4rem;
  height: 4rem;
  padding: 0.8rem 3.2rem;
}

.c0:hover {
  background: #FFBA00;
}

<button
  class="c0"
>
  <span
    class=""
  >
    Button
  </span>
</button>
`;
